### DISCUSSION TITLE
Developers: Which reader has the best KOReader integration?

### DISCUSSION THREAD

**Discussion by zmixnyc** [koreader/koreader#11962] [General]
I have seen some great e-readers coming out and as much as I love my Pocketbook ERA it seems to have some issues with highlighting text consistently.

I figured that any e-reader I might get will be fine as I will only ever use KOReader on it.

So... What's the best hardware these days?

**Comment by mergen3107** [comment]
There were similar discussions here, please search.

You can use tags like "bug" and a platform ("Kobo", "Kindle", "Android", or others) and see what tickets are still open. 

Also on MobileRead: https://www.mobileread.com/forums/showthread.php?t=347826

From my observations for recent new hardware, new Kobos (Clara BW/Color, Libra Color) are quite good - because they have MediaTek processors.

Clara BW is probably a sweet-spot: 1 GHz single-core MediaTek + 512 MB RAM. However, I don't know if there kernel/platform related quirks.

Clara Color has 2 GHz dual-core MediaTek + 1 GB RAM, but it needs to drive both 300 ppi content and 150 ppi color content, so I don't really know how much faster/slower it is against Clara BW when compared on similar things.

Other Kobos from MK7 are quite good as well (maybe except only Libra 2 revisions with quirks). In my own experience, Kobo Clara HD is excellent. 

From my own experience again, I know for a fact that Kindle Scribe is great (1 GHz MediaTek + 1 GB of RAM), barring you can get one with low enough firmware (≤5.16.2.1.1) to get JailBreak on it.
This is my everyday device and I love reading books, manga in CBZ, and even scientific articles without zoom (downloaded straight from my Zotero account) on its gorgeous 300 ppi 10 inch screen.

Android is probably the worst of all, first reason because Android frame buffer handling add much more delay in drawing eink frames. Other reason include but are not limited to bad/terrible OEM implementation that might prevent or impede with your experience, like light controls, etc.

However, in my own experience again, besides that slowness of the screen (which I only realized only after comparing it to Kindle Scribe directly for the first time) Onyx Boox Kon-Tiki 2 KOReader experience was great. I fixed a thing or two while I had device (exporter plugin and light control), so it is even better now :D

You want to avoid older Onyx Boox devices with RockChip CPUs and go for Qualcomm CPUs for many reasons (performance, battery life, and root availability - but that one is beyond the topic of KOReader since you don't need root for pleasant KOReader experience).
I saw good feedback from Onyx Boox Leaf 1/2, Onyx Boox Page users. Maybe others on similar platforms.

I can't say anything about PocketBooks just because I hadn't one, but I know there is a lot of interest and activity about it here on GitHub, many of which were successfully solved (upon my quick throwback in my memory).

**Comment by zmixnyc** [comment]
Thanks for the reply.  I did do  search here before I posted, so sorry,  I must have missed something.

I saw the new Onyx Book Palma today and pretty much hated the built in reader software, but liked the small(ish) form factor and the screen looked great.

Those of you who have tested KOReader on multiple platforms have you been happy with the consistency?

**Reply by Galunid** [reply]
I used it in older kindles, Kobo Clara HD, Kobo Libra H2O, and Boox Nova 2. I liked it the most on Kobo Clara HD. On boox it feels kind of slow, but unless you skip multiple pages/use menu, you won't notice it too much. Kindle was limited because amazon's soft was running in the background all the time. It had less free ram, and was pinging amazon often enough that wifi didn't turn off automatically...

I'd say my tier list is:
1. Kobo Clara HD/Kobo Libra H2O - never had an issue
2. Onyx Boox Nova 2 - feels a bit slow when browsing menu, skipping multiple pages, you have to switch wifi using top menu
3. Kindle Paperwhite 3/4 (?) - Sometimes crashed due to not enough free ram, wifi had to be manually toggled off, *Note: I don't remember which one it was and I don't have it anymore*

I'm not sure about the new Kobo though.

**Reply by Commodore64user** [reply]
> Kindle Paperwhite 3/4 (?) - Sometimes crashed due to not enough free ram

my kindle voyage and k4 don’t seem to suffer from this… (touch wood)

**Reply by Frenzie** [reply]
Depends on the types of documents you open. Outside of heavy PDFs it's generally not an issue.

**Reply by Commodore64user** [reply]
I thought PDFs would use less memory since all the computation has already been done. But then again, you can indeed get very large PDF files unheard of in the ePub world (unless is one of those ridiculous omnibuses with like 20 novels in them)

**Reply by mergen3107** [reply]
PDFs with large uncompressed images are usually an issue.

PDFs with mostly text are OK.

**Reply by Frenzie** [reply]
With PDF the size of the document doesn't matter much, but the (resulting) size of an individual page does. There's an issue around here somewhere of someone with a page that allocates 200+ MB to render.

There is however a strong correlation between the two.

**Comment by pazos** [comment]
> So... What's the best hardware these days?

1. A recent Kobo that has no issues.
2. A recent Pocketbook that has no known issues??.
3. A recent Android device that is supported and has no known issues?????.

You can also repurpose a kindle if you can jailbreak it.

> I will only ever use KOReader on it.

Pocketbook and Android can open links in a web browser.
Pocketbook home screen provides a library view for KOReader.
Android supports 3rd party dictionaries and sharing text with other apps. 

Kobo/Kindle/Remarkable have none of those. If they're not relevant to you go for a kobo. The one that @NiLuJe vouches for ;)

**Reply by NiLuJe** [reply]
So far nothing horrible has happened with the new MTK lineup (besides 5 GHz WiFi being mostly broken). But color sucks (and, more importantly, makes no sense at those screen sizes in my book. I wouldn't be so hostile about it in an Elipsa, for instance), so my advice is still to wait if the Clara B&W's form-factor isn't your thing.

Or you actually really, really, really like the fact that you could display pastel crap, in which case the Libra Colour could make sense, I guess?

**Reply by Commodore64user** [reply]
I read a comment once where you say you were a "kindle guy" what happened? The mobileread threat even has you saying "quick get a kindle pw5". 

Speaking of PW5, i have one but can't JB as is on newer firmware. EVERYONE LOOK THE OTHER WAY NOW, any news on that front?

**Reply by Frenzie** [reply]
> I read a comment once where you say you were a "kindle guy" what happened? The mobileread threat even has you saying "quick get a kindle pw5".

That arguably sounds very self-explanatory. ;-) (Because the only reason to get one ASAP would be if they changed something for the worse.)

**Reply by mergen3107** [reply]
Not a PDF, but the same muPDF engine trying to open this huge vectorized file:

https://github.com/koreader/koreader/issues/11386

**Reply by Commodore64user** [reply]
> Not a PDF, but the same muPDF engine trying to open this huge vectorized file:
> 
> #11386

should this have been in the other comment thread?

**Reply by mergen3107** [reply]
Nope, it is related to the large unwieldy PDFs :D

**Reply by Commodore64user** [reply]
But this is the one about why @NiLuJe is no longer a kindle evangelist not the PDF one.

**Reply by mergen3107** [reply]
Ah, yeah, sorry :D 
I though "thread" is a ticket, but I forgot how awful Questions are

**Reply by NiLuJe** [reply]
> they changed something for the worse.)

Spoiler alert: they did ;p.

> any news on that front?

Something might be coming up Soon(TM).

But depending on how up-to-date your device is, that may not do you much good for now, because of the above (i.e., armhf).

> why @NiLuJe is no longer a kindle evangelist

My stance hasn't changed much: the hardware is very good, everything else is a massive PITA.

**Reply by Commodore64user** [reply]
> Something might be coming up Soon(TM).

Eyes shine brightly

> But depending on how up-to-date your device is, that may not do you much good for now, because of the above (i.e., armhf).

I'm f***ed. 5.16.8 what can i say, I'm a serial updater.

> everything else is a massive PITA.

Luckily my KV does not get new updates so one thing less to worry about 

> > they changed something for the worse.)

Is that from everyone's perspective or just from a "modding one"?

**Reply by Commodore64user** [reply]
In a world where there is no JB available past 5.16.2 (and assuming there is not going to be one soon) what does this mean #11986 for kindle users? What if we assume that there is a JB due soon?

btw 5.16.9 is out now

**Reply by mergen3107** [reply]
Technically, when you have JB and accidentally get updated to > 5.16.2, JB still work (which allows you to run custom commands to downgrade). But that's it for now. 

With at least KOReader working, there is hope that Kindles are useful again :D (but I think gandalf - JB permission binary/script needs to be recompiled too).

For non-Scribe users, I don't think there is much difference between FW versions.

For Scribe users, they can finally keep up with notebook updates while still having KOReader.

**Reply by Commodore64user** [reply]
I see, thanks!

**Comment by dmalinovsky** [comment]
I'm a happy user of KOReader on Pocketbook Inkpad Color 3 (and I like its crappy pastel colors :). 

After some color rendering related fixes I've made, it's working really well on it, performance wise it's similar to the build in reader, and it has pretty good integration — at least I didn't find anything lacking. Also, [Pocketbook Sync plugin](https://github.com/ckilb/pocketbooksync.koplugin) adds nice bits for the home screen integration.

**Reply by NiLuJe** [reply]
> performance wise it's similar to the build in reader

Which is terrible compared to a Kindle or a Kobo, mind you ;).

It might not be terrible enough to bother you, but it *is* a notable difference ;).

**Reply by dmalinovsky** [reply]
Well, maybe it's really visible in the library view, but the reader itself works pretty well. I also own a Kindle Oasis, so I can compare. Of course, the latter is more polished UI and hardware wise, but you have to keep jailbreaking it to have KOReader — which is a big disadvantage to me.

**Reply by NiLuJe** [reply]
The InkView loop is tick-based, with a really awful resolution, so it starts with a pretty severe (self-inflicted) handicap.

Like I said, you might not be able to notice it and/or it might not bother you, but it's there.

**Reply by dmalinovsky** [reply]
Oh, interesting, I didn't know that. I guess my perception isn't sharp enough. :)

**Reply by Frenzie** [reply]
@NiLuJe 
> The InkView loop is tick-based, with a really awful resolution, so it starts with a pretty severe (self-inflicted) handicap.

I presume that explains what @Biep wrote here? https://github.com/koreader/koreader/issues/11668#issuecomment-2158838994

**Reply by NiLuJe** [reply]
> I presume that explains what @Biep wrote here? [#11668 (comment)](https://github.com/koreader/koreader/issues/11668#issuecomment-2158838994)

It doesn't help; but the main explanation for that issue probably lies in the lack of timerfd support for the handling of hold timers, which means we fall-back to a userland timer, and, worse, possibly one based on the time of input *processing* on platforms with (kernel) timestamps we can't trust and/or interpret properly.

That's definitely the case on PB, where we use a `gettimeofday` we compute ourselves when we catch an IV pointer event (and that's the tick-based loop with a crap resolution of something like 50ms I was talking about earlier).

On Android, we don't do timerfds, but we *can* at least trust the kernel timestamp if it looks sane (I don't actually recall when that might not be the case, though, but it should be pretty obvious from looking at verbose debug logs: on a sane (Android) device, you should see different timestamps for each individual event, instead of one per input frame).

**Reply by Frenzie** [reply]
Thanks, I didn't know that about PB
